golos.social/golos.social.cpp

```diff
@@ -113,5 +116,44 @@ void social::unblock(account_name blocker, account_name blocking) {
     });
 }
 
+void social::setreput(account_name voter, account_name author, base_t subtract, base_t add) {
+    tables::reputation_table voter_table(_self, voter);
+    auto voter_rep = voter_table.begin();
+
+    tables::reputation_table author_table(_self, author);
+    auto author_rep = author_table.begin();
+    bool item_exists = (author_rep != author_table.end());
+
+    // Rule #1: Must have non-negative reputation to affect another user's reputation
+    if (voter_rep != voter_table.end() && voter_rep->reputation < 0) {
+        return;
+    }
+
+    if (item_exists) {
+        // Rule #2: If you are downvoting another user, you must have more reputation than him
+        if (add < 0 && !(voter_rep != voter_table.end() && voter_rep->reputation > author_rep->reputation)) {
+            return;
+        }
+
+        author_table.modify(author_rep, 0, [&](auto& item){
+            item.reputation -= subtract;
+            item.reputation += add;
+        });
+
+        return;
+    }
+
+    // Rule #2: If you are downvoting another user, you must have more reputation than him
+    // Author rep is 0 (item not exists), you must have positive rep
+    if (add < 0 && !(voter_rep != voter_table.end() && voter_rep->reputation > 0)) {
+        return;
+    }
+
+    author_table.emplace(author, [&](auto& item){
+        item.account = author;
```

Comment:
All records in this table has `item.account` equals to `author`. For such records you can use eosio::singleton (see: https://github.com/GolosChain/golos-smart/blob/1f40c24c183760673d2873e466ffe2e71b55c4a8/golos.emit/include/golos.emit/golos.emit.hpp#L19)